src/async_fuse/memfs/open_file.rs

```diff
@@ -0,0 +1,127 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use parking_lot::{Mutex, RwLock};
+
+use super::fs_util::FileAttr;
+use crate::async_fuse::fuse::protocol::INum;
+
+/// A structure representing an open file with its attributes and open count.
+///
+/// The `attr` field contains the file attributes, while `open_cnt` keeps track
+/// of the number of times this file is currently opened.
+#[derive(Debug)]
+#[allow(clippy::partial_pub_fields)] // Pub attr for simplicity.
+pub struct RawOpenFile {
+    /// The file attributes.
+    pub attr: FileAttr,
+    /// The number of times this file is currently opened.
+    open_cnt: u32,
+}
+
+/// A thread-safe reference counted wrapper around `RawOpenFile`.
+///
+/// This type uses `Arc` for shared ownership and `RwLock` for thread-safe
+/// mutability, allowing multiple readers or one writer at the same time.
+pub type OpenFile = Arc<RwLock<RawOpenFile>>;
+
+/// Internal structure to keep track of all open files.
+///
+/// This structure uses a `HashMap` to associate inode numbers (`INum`) with
+/// their corresponding `OpenFile` instances.
+#[derive(Debug)]
+struct RawOpenFiles {
+    /// The collection of open files.
+    open_files: HashMap<INum, OpenFile>,
+}
+
+/// A thread-safe, mutable collection of open files.
+///
+/// This structure provides an interface to interact with open files,
+/// encapsulating the internal locking mechanism through `Mutex`.
+#[derive(Debug)]
+pub struct OpenFiles {
+    /// The internal collection of open files.
+    inner: Arc<Mutex<RawOpenFiles>>,
+}
+
+impl OpenFiles {
+    /// Constructs a new `OpenFiles` collection.
+    ///
+    /// This initializes an empty collection of open files.
+    pub fn new() -> Self {
+        OpenFiles {
+            inner: Arc::new(Mutex::new(RawOpenFiles {
+                open_files: HashMap::new(),
+            })),
+        }
+    }
+
+    /// Try to retrieves an `OpenFile` by its inode number.
+    ///
+    /// Returns `Some(OpenFile)` if the file is open, or `None` if not found.
+    pub fn try_get(&self, inum: INum) -> Option<OpenFile> {
+        let inner = self.inner.lock();
+        let open_file = inner.open_files.get(&inum).map(Arc::clone);
+        open_file
+    }
+
+    /// Retrieves an `OpenFile` by its inode number.
+    ///
+    /// Panics if the file is not found.
+    #[allow(clippy::unwrap_used)] // We should panic here
+    pub fn get(&self, inum: INum) -> OpenFile {
+        self.try_get(inum)
+            .unwrap_or_else(|| panic!("Couldn't find the open file with inum={inum}"))
+    }
+
+    /// Opens a file, adding it to the collection
+    ///
+    /// Returns a reference to the newly opened file.
+    pub fn open(&self, inum: INum, attr: FileAttr) -> OpenFile {
+        let mut inner = self.inner.lock();
+        let open_file = Arc::new(RwLock::new(RawOpenFile { attr, open_cnt: 1 }));
+        inner.open_files.insert(inum, Arc::clone(&open_file));
+        open_file
+    }
```

Comment:
Do we need to maintain some attributes when a file is opened again? For example, if something changed from the remote nodes and they are flushed to the KV, would them be visible for the local node?

Reply:
No , they just see the local node